driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java

```diff
@@ -214,11 +218,13 @@ public <T> T withTransaction(final TransactionBody<T> transactionBody) {
     public <T> T withTransaction(final TransactionBody<T> transactionBody, final TransactionOptions options) {
         notNull("transactionBody", transactionBody);
         long startTime = ClientSessionClock.INSTANCE.now();
+        TimeoutContext withTransactionTimeoutContext = createTimeoutContext(options);
+
         outer:
         while (true) {
             T retVal;
             try {
-                startTransaction(options);
+                startTransaction(options, copyTimeoutContext(withTransactionTimeoutContext));
```

Comment:
This ensures that on failure (which issues an abort transaction) that the original `withTransactionTimeoutContext` isn't reset.